Implementing Issue #408

MHaselmaier opened the pull request:
I gave it a shot at implementing issue #408.

I wasn't sure if the timeout parameter should be added to the get-function of core::provider::Provider or core::provider::PoolTemplate. I settled for the latter.

I think I found a bug while writing the test:
PoolTemplate::createShared calls PoolTemplate::startCleanupTask, which takes a parameter.
Apperantly this function PoolTemplate::createShared was never called and since PoolTemplate is a template no compiler errors occurred. I fixed it by passing ptr to PoolTemplate::startCleanupTask.

MHaselmaier commented:
Seems like my test only runs on windows. 
I suspect the implementation of std::async. I only have access to a windows system but I will try to investigate it further.

MHaselmaier commented:
As far as I can see the code should also work on linux.
I have an example here on compiler explorer: https://godbolt.org/z/Garaabz4b

@lganzzzo Do you have an idea?

lganzzzo reviewed (changes requested):
Looks great!

> Seems like my test only runs on windows.
I suspect the implementation of std::async. I only have access to a windows system but I will try to investigate it further.

As I've checked on Mac the problem is caused by the timeout = std::chrono::microseconds::max().
Turns out on Mac timeout = std::chrono::microseconds::max() causes m_condition.wait_for() to return immediately.

bamkrs commented on src/oatpp/core/provider/Pool.hpp:
```diff
@@ -223,22 +223,24 @@ class PoolTemplate : public oatpp::base::Countable, public async::CoroutineWaitL
   }
 
-  std::shared_ptr<TResource> get(const std::shared_ptr<PoolTemplate>& _this) {
+  std::shared_ptr<TResource> get(const std::shared_ptr<PoolTemplate>& _this, const std::chrono::duration<v_int64, std::micro>& timeout = std::chrono::microseconds::max()) {
```
Its not the problem of being `::max()`. In `libstdc++` this behavior is by design. 

> an object of type std::chrono::duration representing the maximum time to spend waiting. Note that rel_time must be small enough not to overflow when added to std::chrono::steady_clock::now().

And looking at the source of `libstdc++` we can see why:

```c++

template<typename _Rep, typename _Period>
  cv_status
  wait_for(unique_lock<mutex>& __lock, 
                const chrono::duration<_Rep, _Period>& __rtime)
               { return wait_until(__lock, __clock_t::now() + __rtime); }
```

So I assume we just "overflow" when passing `::max()`. 
Therefore we could either catch `::max()` in an if or directly go for `wait_until` and do our own calculation with overflow protection. The last option would also catch other conditions where the passed timeout would overflow.

MHaselmaier replied:
Thanks for the feedback, I'll implement it later today.

I thoght a bit about a way to implement it a different way (without an default value).
Since we most likely have to add extra logic, why don't we provide to public get() functions (one with timeout, one wihtout),
aquire the lock and pass it to an private get function which handles the resource acquisition?
Might be a bit cleaner to read.

MHaselmaier replied:
> However, I think it would also be sufficient to check for `::max()` since I can not imagine an use-case where values greater than a couple of (milli)seconds or `::max()` would be passed.

Yes that's what I thought as well. In that case it would be easy to just have to public get funcitons.

bamkrs replied:
> I thoght a bit about a way to implement it a different way (without an default value).
> Since we most likely have to add extra logic, why don't we provide to public get() functions (one with timeout, one wihtout),

*performance wise*  this is the nicer solution. 

> > However, I think it would also be sufficient to check for ::max() since I can not imagine an use-case where values greater than a couple of (milli)seconds or ::max() would be passed.
>
> Yes that's what I thought as well. In that case it would be easy to just have to public get funcitons.

I would wait for @lganzzzo thoughts on that. But I really think this would be sufficient and we could add a warning to the code-doc to not overflow the timeout.

lganzzzo replied:
I agree overloading might be a better solution than branching with `if`s.

> But I really think this would be sufficient and we could add a warning to the code-doc to not overflow the timeout.

Yes, code-doc is always a good idea.
Also, think we shouldn't worry too much for cases when the user explicitly calls get with `::max()` argument.

The end-user should have the following APIs:

```cpp
/* when there is no resource available */

pool.get();      // blocks forever 
pool.getAsync(); // blocks forever

pool.get(<timeout>);      // blocks until timeout 
pool.getAsync(<timeout>); // blocks until timeout
```

MHaselmaier replied:
Are you speaking about the Pool-functions or the PoolTemplate-functions?
In the other conversation you said:
> Yes, we can make `get(const std::shared_ptr<PoolTemplate>& _this)` and `getAsync(const std::shared_ptr<PoolTemplate>& _this)` static and protected.
> Users won't be affected since they are not supposed to use those functions directly.
> Instead user should use `get()` and `getAsync()`.

The Pool-functions are inherited from core::provider::Provider.
Should we then add new pure virutal functions with timeout parameters in the Provider class, or just add the new overloads to the Pool class. 

In the library scop there would be the following other classes to extend:

- network::virutal_::server::ConnectionProvider
- network::virutal_::client::ConnectionProvider
- network::tcp::server::ConnectionProvider
- network::tcp::client::ConnectionProvider

lganzzzo replied:
> Are you speaking about the Pool-functions or the PoolTemplate-functions?
In the other conversation you said:

Overloading vs ifs - it about PoolTemplate.

> The Pool-functions are inherited from core::provider::Provider.
Should we then add new pure virutal functions with timeout parameters in the Provider class, or just add the new overloads to the Pool class.

Yes, the new interface should be applied to all providers.

> In the library scop there would be the following other classes to extend:

This is actually a good point. 
While its understood that we have to update providers interface, I dont like the idea 
of having to separate virtual methods with and without timeout.
So the end user interface should have get and getAsync with timeout and a default value

MHaselmaier replied:
Default parameter values for virtual functions can be very complicated:
https://eel.is/c++draft/dcl.fct.default#10

If that's fine, I'll change the Provider interace and add a new defaulted timeout parameter to get and getAsync.

lganzzzo replied:
Sure, default params should be added to derived classes only.

MHaselmaier commented on src/oatpp/core/provider/Pool.hpp:
```diff
@@ -223,22 +223,24 @@ class PoolTemplate : public oatpp::base::Countable, public async::CoroutineWaitL
-        m_condition.wait(guard);
+      auto finishedPredicate = [this]() { return !m_running || !m_bench.empty() || m_counter < m_maxResources; };
+      if (!m_condition.wait_for(guard, timeout, std::move(finishedPredicate)))
+      {
```
I tried to look out for things like that...
Have you thought about using tools like cland-format for things like that?

lganzzzo replied:
Yes, having cland-format would be nice. 
However, I personally treat it as a super minor thing.

MHaselmaier commented on src/oatpp/core/provider/Pool.hpp:
```diff
@@ -223,22 +223,24 @@ class PoolTemplate : public oatpp::base::Countable, public async::CoroutineWaitL
     /* "new" is called directly to keep constructor private */
     auto ptr = std::shared_ptr<PoolTemplate>(new PoolTemplate(provider, maxResources, maxResourceTTL.count()));
-    ptr->startCleanupTask();
+    ptr->startCleanupTask(ptr);
```
What about get() and getAsync(). 
As far as I can see the get passed a shared_ptr to this so that the AcquisitionProxy keeps the pool alive.
I don't think it is desirable to call both functions with a shared_ptr pointing to another object than this.

Should they be also static? I am not sure how much current users of oat++ will be affected by such a change.
Maybe they would have to make changes to their code to upgrade to a newer version if this change is made.

lganzzzo replied:
Yes, we can make  `get(const std::shared_ptr<PoolTemplate>& _this)` and `getAsync(const std::shared_ptr<PoolTemplate>& _this)` static and protected.
Users won't be affected since they are not supposed to use those functions directly.
Instead user should use `get()` and `getAsync()`.

MHaselmaier commented:
I added the optional parameter to Provider::get and Provider::getAsync.

Since the concrete Provider-Instances are used through an pointer to the abstract Provider-classes I added the default value to every function declaration, not just only in the derived classes.
Otherwise we would have to provide an timeout explicitly everywere.

I opted for std::chrono::microseconds::zero() as the default because the zero values should be convertible without overflow between the different representations.

So far I only implemented the timeout on the ::get functions.
For now I haven't decided how to implement it inside the coroutines.

lganzzzo commented:
Looks good,
I'll review it in more detail later today.
Also, we have to find out why CI reported a memory leak in `PoolTemplateTest` on Windows and Ubuntu 18.04

MHaselmaier commented:
Yes I have seen the failing tests. Locally I was not able to reproduce it. I suppect some object is being kept alive by a shared_ptr in a detached thread. Maybe some debug log infos would help to narrow down the failure.

lganzzzo commented on src/oatpp/network/tcp/client/ConnectionProvider.cpp:
```diff
@@ -124,32 +125,84 @@ std::shared_ptr<oatpp::data::stream::IOStream> ConnectionProvider::get() {
+        return {false, future.get()};
+      }
+
+      std::thread{std::move(task)}.detach();
```
Unfortunately, we can't spawn threads from coroutines. An application can spawn millions of coroutines
but the max threads-count is very limited. If each coroutine spawns a thread - an application will crash.

lganzzzo commented on src/oatpp/network/tcp/client/ConnectionProvider.cpp:
```diff
@@ -252,7 +308,9 @@ oatpp::async::CoroutineStarterForResult<const std::shared_ptr<oatpp::data::strea
         return _return(std::make_shared<oatpp::network::tcp::Connection>(m_clientHandle));
       }
-      if(error == WSAEWOULDBLOCK || error == WSAEINPROGRESS) {
+      if (timedout()) {
```
This won't work because of the following:

```cpp 
if(error == WSAEWOULDBLOCK || error == WSAEINPROGRESS) {
  return ioWait(m_clientHandle, oatpp::async::Action::IOEventType::IO_EVENT_WRITE);
}
```

This action says to the coroutine processor - wait until the socket is available for writing.
And there is no timeout here. It can potentially last forever.

lganzzzo commented on src/oatpp/network/virtual_/client/ConnectionProvider.cpp:
```diff
@@ -43,44 +45,68 @@ void ConnectionProvider::stop() {
+    return future.get();
+  }
+
+  std::thread{std::move(task), m_interface, m_maxAvailableToRead, m_maxAvailableToWrite}.detach();
```
We don't want to spawn threads in `ConnectionProvider::get` since it's too resource-inefficient.

lganzzzo commented on src/oatpp/network/virtual_/server/ConnectionProvider.cpp:
```diff
@@ -52,12 +53,23 @@ void ConnectionProvider::stop() {
-  return socket;
+
+  std::thread{std::move(task)}.detach();
+  return future.wait_for(timeout) == std::future_status::ready ? future.get() : nullptr;
```
We don't want to spawn threads in ConnectionProvider::get since it's too resource-inefficient.

lganzzzo reviewed (changes requested):
This is a great work since it reveals the challenges of implementing the following APIs:

```cpp
Provider::get(<timeout>);
Provider::getAsync(<timeout>);
```

Given those challenges, I see the following cons of the above API:
- It makes implementing custom providers too complicated.
- In some cases, it may introduce unwanted additional latency obtaining a resource.

My new proposition would be the following:

- return to the old provider API:
   ```cpp
   Provider::get();
   Provider::getAsync();
   ```
- Implement timeout (for now) only for Pool.
- Configure timeout via Pool constructor parameter.

MHaselmaier commented:
Thanks for the feedback. I had similar thoughts in the back off my mind and I'm glad you see it that way too.
I'll implement the requested changes.

MHaselmaier commented on src/oatpp/core/provider/Pool.hpp:
```diff
@@ -210,79 +237,64 @@ class PoolTemplate : public oatpp::base::Countable, public async::CoroutineWaitL
+          std::unique_lock<std::mutex> guard(m_pool->m_lock, std::try_to_lock);
+
+          if (!guard.owns_lock()) {
+            return timedout() ? this->_return(nullptr) : this->waitRepeat(std::chrono::milliseconds(100));
```
I had a look at `CoroutineWaitList` and the `Action` Union-Member `Data` and was wondering how we would pass the timeout. 
Would it be ok to create a new struct which incapsulates the `CoroutineWaitList*` and a `std::duration::steady_clock::timep_point` to represent the time the timeout would be exceeded?

lganzzzo replied:
Hey @MHaselmaier ,

Sorry, just found your comment - I guess I missed it during the previous review.

> I had a look at CoroutineWaitList and the Action Union-Member Data and was wondering how we would pass the timeout.

We want to keep `Action` as small as possible and for this reason, we don't want to put big structures in the union. 
However, since `union Data`  already has a `FunctionPtr` member which takes 16 bytes - we have plenty of place to put there a pointer to WaitList + timeout.

```cpp
  struct WaitListWithTimeout {
    CoroutineWaitList* waitList;
    v_uint64 timeout; // or std::chrono::steady_clock::time_point timeout
  };

  union Data {
    FunctionPtr fptr;
    AbstractCoroutine* coroutine;
    Error* error;
    IOData ioData;
    v_int64 timePointMicroseconds;
    WaitListWithTimeout waitListWithTimeout;
  };
```

MHaselmaier commented on src/oatpp/core/provider/Pool.hpp:
```diff
@@ -189,100 +190,91 @@ class PoolTemplate : public oatpp::base::Countable, public async::CoroutineWaitL
+
     try {
-      return std::make_shared<AcquisitionProxyImpl>(m_provider->get(), _this);
+      return std::make_shared<AcquisitionProxyImpl>(_this->m_provider->get(), _this);
```
I was thinking: Since `_this->m_provider->get()` may block, don't we need a timeout here as well?

lganzzzo replied:
As the original issue states - we want to implement a timeout for pool resource acquisition.
When there are no free resources available in the pool we want to be able to limit our waiting time.

On one hand, at this particular point, the resource is not yet allocated but on the other, our request for the resource is already being processed. So it's kinda a gray zone.

I propose to treat pool timeout as a "pool resource acquisition timeout". So there is no need for the provider's `get()` timeout since the pool is already allocated a slot for us.

Also, if we introduce timeout in `providers->get()` again - we'll return to where we began from:)

lganzzzo commented on src/oatpp/core/provider/Pool.hpp:
```diff
@@ -305,7 +297,7 @@ class PoolTemplate : public oatpp::base::Countable, public async::CoroutineWaitL
         }
 
-        return m_provider->getAsync().callbackTo(&GetCoroutine::onGet);
+        return m_pool->m_provider->getAsync().callbackTo(&GetCoroutine::onGet);
```
Please see my comment for `m_pool->m_provider->get()`

lganzzzo commented:
Hey @MHaselmaier ,

Please feel free to join [Oat++ devs group on Gitter](https://gitter.im/oatpp-framework/Lobby) - you may ping me there in case I'm not responding for too long.

lganzzzo commented on src/oatpp/core/async/CoroutineWaitList.cpp:
```diff
@@ -25,16 +26,69 @@
   notifyAll();
+  m_stop = true;
+  if (m_thread.joinable()) {
+    m_thread.join();
```
This can potentially cause a performance issue.

```cpp
class MyCoroutine ... {
  CoroutineWaitList m_waitList;
}
```

Such coroutine will block the async processor for `100ms` when destroyed.

lganzzzo commented on src/oatpp/core/async/CoroutineWaitList.cpp:
```diff
@@ -25,16 +26,69 @@
+}
+
+void CoroutineWaitList::startTimeoutCheckerThread() {
+  m_thread = std::thread{[this]() { checkCoroutinesForTimeouts(); }};
```
Since we don't want to join thread during coroutine destruction I would propose to move this thread to `async::Processor`.
- This thread will serve all the waitlists.
- This thread will call `checkCoroutinesForTimeouts(<processor>)` every `100ms`

lganzzzo reviewed (commented):
Looks good. However, we have to think of a workaround in order not
to join threads in the `CoroutineWaitList` destructor.
Plus, it's better not to spawn threads from primitives that can be used in Coroutines.

lganzzzo commented:
/AzurePipelines run

MHaselmaier commented:
/AzurePipelines run

lganzzzo commented on src/oatpp/core/async/Processor.hpp:
```diff
@@ -108,8 +112,19 @@ class Processor {
+
+private:
+
+  std::recursive_mutex m_coroutineWaitListsWithTimeoutsMutex;
```
Do we need this mutex to be recursive?
At a first glance, it seems we can have it as a simple `std::mutex`.

MHaselmaier replied:
I made it a recursive_mutex, because in `Processor::checkCoroutinesForTimeouts()` we call `CoroutineWaitList::checkCoroutinesForTimeouts()` which in return could remove the current WaitList from the list of coroutinewaitlists with timeout from the processor. Becaus we need to hold the lock of `m_coroutineWaitListsWithTimeoutsMutex` while we are inside `Processor::checkCoroutinesForTimeouts()`  it has to be recursive. 

In the destructor of `CoroutineWaitList` we would remove the waitlist from the processors list as well. I played with the idea of having a second list with "to-remove-waitlists" and at the begining of `Processor::checkCoroutinesForTimeouts()` work through this list and merge it with the actual `m_coroutineWaitListsWithTimeouts`. But this would introduce a race condition where the `CoroutineWaitList` is already destroy while the `Processors` still holds a pointer to the now deleted object and calls the `CoroutineWaitList::checkCoroutinesForTimeouts()` method.

lganzzzo commented on src/oatpp/core/async/Processor.hpp:
```diff
@@ -123,10 +138,13 @@ class Processor {
-  {}
+  Processor() = default;
+
+  ~Processor() {
```
Let's move this logic to `Processor::stop()` method

MHaselmaier replied:
Keep in mind: If we move this logic to `Processor::stop()` and call the destructor without calling the `Processor::stop()` we get a call to std::terminate.

How about joining the thread in stop, but as a fallback also join it in the Destructor if the thread is still joinable?

lganzzzo replied:
Currently `stop()` is always called before destruction.
Processor is used via `Executor::SubmissionProcessor` wrapper - which is possible another thing to refactor (but it's out of the scope of this PR).

So it's better to have all the joins in the `stop()` method - to follow the existing pattern.

> Keep in mind: If we move this logic to Processor::stop() and call the destructor without calling the Processor::stop() we get a call to std::terminate.

I'm ok with that. In this case `std::terminate` will indicate the API misuse.

lganzzzo reviewed (commented):
Looks good.
Please see minor comments

lganzzzo commented:
Great Work!
Merged
